src/MinimiseProgramTransformer.cpp

```diff
@@ -0,0 +1,339 @@
+
+    bool validMapping = true;
+    for (size_t i = 0; i < leftAtoms.size() && validMapping; i++) {
+        // match arguments
+        std::vector<AstArgument*> leftArgs = leftAtoms[i]->getAtom()->getArguments();
+        std::vector<AstArgument*> rightArgs = rightAtoms[i]->getAtom()->getArguments();
+
+        for (size_t j = 0; j < leftArgs.size(); j++) {
+            AstArgument* leftArg = leftArgs[j];
+            AstArgument* rightArg = rightArgs[j];
+            if (isVariable(leftArg) && isVariable(rightArg)) {
+                // both variables, their names should match to each other through the clause
+                std::string leftVarName = dynamic_cast<AstVariable*>(leftArg)->getName();
+                std::string rightVarName = dynamic_cast<AstVariable*>(rightArg)->getName();
+
+                std::string currentMap = variableMap[leftVarName];
+                if (currentMap == "") {
+                    // unassigned yet, so assign it appropriately
+                    variableMap[leftVarName] = rightVarName;
+                } else if (currentMap != rightVarName) {
+                    // mapping is inconsistent!
+                    // clauses cannot be equivalent under this permutation
+                    validMapping = false;
+                    break;
+                }
+            } else if (isConstant(leftArg) && isConstant(rightArg)) {
+                // check if its the same constant
+                auto leftCst = dynamic_cast<AstConstant*>(leftArg)->getIndex();
+                auto rightCst = dynamic_cast<AstConstant*>(rightArg)->getIndex();
+
+                if (leftCst != rightCst) {
+                    // constants don't match, failed!
+                    validMapping = false;
+                    break;
+                }
+            } else {
+                // not the same type, failed!
+                validMapping = false;
+                break;
+            }
+        }
+    }
+
+    // return whether a valid variable mapping exists for this permutation
+    return validMapping;
+}
+
+/**
+ * Check whether two clauses are bijectively equivalent.
+ */
+bool areBijectivelyEquivalent(AstClause* left, AstClause* right) {
+    // only check bijective equivalence for a subset of the possible clauses
+    auto isValidClause = [&](const AstClause* clause) {
+        bool valid = true;
+
+        // check that all body literals are atoms
+        // i.e. avoid clauses with constraints or negations
+        for (AstLiteral* lit : clause->getBodyLiterals()) {
+            if (!dynamic_cast<AstAtom*>(lit)) {
+                valid = false;
+            }
+        }
+
+        // check that all arguments are either constants or variables
+        // i.e. only allow primitive arguments
+        visitDepthFirst(*clause, [&](const AstArgument& arg) {
+            if (!dynamic_cast<const AstVariable*>(&arg) && !dynamic_cast<const AstConstant*>(&arg)) {
+                valid = false;
```

Comment:
You could avoid needing the boolean here if your lambda threw an exception, but I don't think it would be worth the extra code complexity if you only have a few arguments.